Src/PCompiler/CompilerCore/Backend/Java/JavaCodeGenerator.cs

```diff
@@ -220,17 +220,52 @@ private void WriteEventDecl(PEvent e)
             string eventName = _context.Names.GetNameForDecl(e);
 
             TypeManager.JType argType = _context.Types.JavaTypeFor(e.PayloadType);
-            switch (argType)
+            bool hasPayload = !(argType is TypeManager.JType.JVoid);
+
+            WriteLine($"public static class {eventName} implements PObserveEvent.PEvent<{argType.ReferenceTypeName}> {{");
+
+            if (hasPayload)
             {
-                case TypeManager.JType.JVoid _:
-                    // Special-case an event with no payload: just emit an empty record.
-                    WriteLine($"public record {eventName}() implements PObserveEvent.PEvent {{ }} ");
-                    break;
-                default:
-                    WriteLine($"public record {eventName}({argType.TypeName} payload) implements PObserveEvent.PEvent {{ }} ");
-                    break;
+                WriteLine($"public {eventName}({argType.TypeName} p) {{ this.payload = p; }}");
+            }
+            else
+            {
+                WriteLine($"public {eventName}() {{ }}");
             }
 
+            WriteLine($"private {argType.ReferenceTypeName} payload; ");
+            WriteLine($"public {argType.ReferenceTypeName} getPayload() {{ return payload; }}");
+            WriteLine();
+
+            WriteLine("@Override");
+            WriteLine("public boolean equals(Object o) {");
```

Comment:
Can this by any chance be moved into a base class? Instead of generating it for each event?

Reply:
I meant the generic code like `hashCode` and `equals`

Reply:
I don't think so - how would the base class look?  We need to do a field-by-field comparison, and the base class would have no way of iterating over those fields (short of something gnarly like using reflection, which is best avoided here I think)

Reply:
That said, though, we _could_ simplify `equals` by calling directly into `Values.deepEquals()`, since all those checks are already done on the prt side.

Reply:
I like both options, anything that reduces the generated code is great!

Reply:
Yep, thanks for this observation!  After [this change](https://github.com/dijkstracula/prtsandbox/commit/4ad482a11fdce75d60f11d463c37afa6d0348cb9) in the runtime, we don't need to emit that boilerplate over and over again.